packages/flutter_tools/lib/src/commands/analyze.dart

```diff
@@ -457,4 +474,268 @@ linter:
+
+  void _sendCommand(String method, Map<String, dynamic> params) {
+    String message = JSON.encode(<String, dynamic> {
+      'id': (++_id).toString(),
+      'method': method,
+      'params': params
+    });
+    _process.stdin.writeln(message);
+    printTrace('==> $message');
+  }
+
+  void _handleServerResponse(String line) {
+    printTrace('<== $line');
+
+    dynamic response = JSON.decode(line);
+
+    if (response is Map) {
+      if (response['event'] != null) {
+        String event = response['event'];
+        dynamic params = response['params'];
+
+        if (params is Map) {
+          if (event == 'server.status')
+            _handleStatus(response['params']);
+          else if (event == 'analysis.errors')
+            _handleAnalysisIssues(response['params']);
+          else if (event == 'server.error')
+            _handleServerError(response['params']);
+        }
+      } else if (response['error'] != null) {
+        printError('Error from the analysis server: ${response['error']['message']}');
+      }
+    }
+  }
+
+  void _handleStatus(Map<String, dynamic> statusInfo) {
+    // {"event":"server.status","params":{"analysis":{"isAnalyzing":true}}}
+    if (statusInfo['analysis'] != null) {
+      bool isAnalyzing = statusInfo['analysis']['isAnalyzing'];
+      _analyzingController.add(isAnalyzing);
+    }
+  }
+
+  void _handleServerError(Map<String, dynamic> errorInfo) {
+    printError('Error from the analysis server: ${errorInfo['message']}');
+  }
+
+  void _handleAnalysisIssues(Map<String, dynamic> issueInfo) {
+    // {"event":"analysis.errors","params":{"file":"/Users/.../lib/main.dart","errors":[]}}
+    String file = issueInfo['file'];
+    List<AnalysisError> errors = issueInfo['errors'].map((Map<String, dynamic> json) => new AnalysisError(json)).toList();
+    _errorsController.add(new FileAnalysisErrors(file, errors));
+  }
+
+  Future dispose() async => _process?.kill();
+}
+
+class FileAnalysisErrors {
+  FileAnalysisErrors(this.file, this.errors);
+
+  final String file;
+  final List<AnalysisError> errors;
+}
+
+class AnalysisError implements Comparable<AnalysisError> {
+  AnalysisError(this.json);
+
+  static final Map<String, int> _severityMap = <String, int> {
+    'ERROR': 3,
+    'WARNING': 2,
+    'INFO': 1
+  };
+
+  // "severity":"INFO","type":"TODO","location":{
+  //   "file":"/Users/.../lib/test.dart","offset":362,"length":72,"startLine":15,"startColumn":4
+  // },"message":"...","hasFix":false}
+  Map<String, dynamic> json;
+
+  String get severity => json['severity'];
+  int get severityLevel => _severityMap[severity] ?? 0;
+  String get type => json['type'];
+  String get message => json['message'];
+
+  String get file => json['location']['file'];
+  int get startLine => json['location']['startLine'];
+  int get startColumn => json['location']['startColumn'];
+
+  int compareTo(AnalysisError other) {
+    // Sort in order of severity, file path, source line, and message.
```

Comment:
As a general rule I agree with you, but in Dart in my experience it's common to have errors be side-effects of earlier errors, where the earlier error is only a warning or even a lint. For example:

```
class A {
  B(bool test) { }
}

void main() {
  new A(true);
}
```

The first message is a lint saying "B" is a bad name for a method ("Name non-constant identifiers using lowerCamelCase."). The actual error here is that I copied and pasted class B, then renamed it to A, and forgot to change the constructor name. So the actual error I need to fix is on the line for this lint, at the exact location pointed to by the lint.

The highest severity error is "0 positional arguments expected, but 1 found", on the constructor call. This doesn't point to the line with the error, and it doesn't really describe anything related to the actual error.